test/integration/finance/api/invoices_test.rb

```diff
@@ -1,210 +1,199 @@
+      assert_response :success
+    end
 
-          get "/api/invoices.xml?access_token=#{token.value}"
+    test 'deny access if member does not have finance permission' do
+      @provider.settings.allow_finance!
+      member = FactoryBot.create(:member, account: @provider, admin_sections: [])
+      token = FactoryBot.create(:access_token, owner: member, scopes: ['finance'])
 
-          assert_response :success
-        end
+      get "/api/invoices.xml?access_token=#{token.value}"
 
-        should 'deny access if member does not have finance permission' do
-          @provider.settings.allow_finance!
-          member = FactoryBot.create(:member, account: @provider, admin_sections: [])
-          token  = FactoryBot.create(:access_token, owner: member, scopes: ['finance'])
+      assert_response :forbidden
+    end
 
-          get "/api/invoices.xml?access_token=#{token.value}"
+    test 'return 404 on non-existent invoice' do
+      get "/api/#{@context}invoices/42_ID.xml?provider_key=#{@key}"
+      assert_response :not_found
+    end
 
-          assert_response :forbidden
-        end
+    test 'return XML invoice specified by ID' do
+      setup_existing_invoices_test
+      get "/api/#{@context}invoices/#{@invoice.id}.xml?provider_key=#{@key}"
+
+      assert_response :success
+      assert_select 'invoice' do
+        assert_select 'invoice > id', text: @invoice.id.to_s
+        assert_select 'invoice > friendly_id', text: @invoice.friendly_id.to_s
+        assert_select 'invoice > buyer > id', @invoice.buyer_account.id.to_s
+        assert_select 'line-items > line-item > cost', '10.0'
+        assert_select 'invoice > cost', '20.0'
+        # vat_rate and vat_amount do not show when vat_rate = 0
+        assert_select 'invoice > vat_rate', text: "0", count: 0
+        assert_select 'invoice > vat_amount', text: "0", count: 0
+        assert_select 'invoice > cost_without_vat', '20.0'
+        assert_select 'invoice > payment_transactions_count', "0"
       end
+    end
 
-      should 'return 404 on non-existent invoice' do
-        get "/api/#{@context}invoices/42_ID.xml?provider_key=#{@key}"
-        assert_response :not_found
+    test 'work fine with deleted account' do
+      setup_existing_invoices_test
+      @buyer = FactoryBot.create(:buyer_account, provider_account: @provider)
+      @invoice = FactoryBot.create(:invoice, provider_account: @provider, buyer_account: @buyer)
+      @buyer.delete # destroy
+
+      get "/api/invoices/#{@invoice.id}.xml?provider_key=#{@key}"
+
+      assert_response :success
+      assert_select 'invoice' do
+        assert_select 'invoice > buyer > id', @buyer.id.to_s
+        assert_select 'invoice > buyer > status', "deleted"
       end
+    end
+
+    test "have vat_rate if non negative" do
+      setup_existing_invoices_test
+      @buyer.vat_rate = 10.0
+      @buyer.save
+
+      get "/api/#{@context}invoices.xml?provider_key=#{@key}"
 
-      context 'with existing invoices' do
```

Comment:
for `      context 'with existing invoices' do` you could create a sub-class to hold the setup method but current approach with `#setup_existing_invoices_test` is reasonable. Just saying.

Reply:
No, you're right, sub-classes is better in this case.